File: sum.pl
sum_list([], 0).        % Base case: The sum of an empty list is 0.
sum_list([X|Xs], Sum) :-  % Recursive case: Sum of the list = X + Sum of the rest of the list.
    sum_list(Xs, Rest),
    Sum is X + Rest.


sum_list([], 0).
sum_list([X|Xs], Sum) :-
    X > 0,
    sum_list(Xs, Rest),
    Sum is X + Rest.
sum_list([X|Xs], Sum) :-
    X <= 0,
    sum_list(Xs, Sum).


sum_elements(List, Sum) :-
    sum_list(List, Sum).


sum_odd_numbers([], 0).
sum_odd_numbers([X|Xs], Sum) :-
    X mod 2 =:= 1, 
    sum_odd_numbers(Xs, Rest),
    Sum is X + Rest.
sum_odd_numbers([X|Xs], Sum) :-
    X mod 2 =:= 0,
    sum_odd_numbers(Xs, Sum).

File: sum.hs
sumOddNumbers :: [Int] -> Int
sumOddNumbers numbers = sum [x | x <- numbers, odd x]

main :: IO ()
main = do
    let myList = [1, 2, 3, 4, 5, 6, 7, 8, 9]
    let result = sumOddNumbers myList
    putStrLn ("Sum of odd numbers: " ++ show result)

productOfOddNumbers :: [Int] -> Int
productOfOddNumbers [] = 1  -- Base case: empty list has a product of 1
productOfOddNumbers (x:xs)
    | odd x     = x * productOfOddNumbers xs  -- Multiply odd numbers
    | otherwise = productOfOddNumbers xs      -- Skip even numbers

main :: IO ()
main = do
    let myList = [1, 2, 3, 4, 5]
    let result = productOfOddNumbers myList
    putStrLn ("Product of odd numbers: " ++ show result)

